src/compiler/checker.ts

```diff
@@ -45494,8 +45494,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                         }
                         if (isConstantVariable(symbol)) {
                             const declaration = symbol.valueDeclaration;
-                            if (declaration && isVariableDeclaration(declaration) && !declaration.type && declaration.initializer && (!location || declaration !== location && isBlockScopedNameDeclaredBeforeUse(declaration, location))) {
-                                return evaluate(declaration.initializer, declaration);
+                            if (declaration && isVariableDeclaration(declaration) && declaration.initializer && (!location || declaration !== location && isBlockScopedNameDeclaredBeforeUse(declaration, location))) {
+                                if (!declaration.type) {
+                                    return evaluate(declaration.initializer, declaration);
+                                }
+                                const declaredType = getTypeOfSymbol(symbol);
+                                const type = declaredType.flags & TypeFlags.Union ?
+                                    getAssignmentReducedType(declaredType as UnionType, getInitialType(declaration)) :
+                                    declaredType;
+                                if (type.flags & TypeFlags.StringOrNumberLiteral) {
+                                    return (type as StringLiteralType | NumberLiteralType).value;

```

Comment:
Note that the initializer is still required for all of this to work here. I'm not sure what's the preferred way of handling potential errors in situations like this:

```ts
const str: 'foo' = 'bar'
const another = `_${str}` // "_foo" or "_bar" ?
```

I think the argument can be made both ways. I'm not sure if this matters all that much when the program errors anyway - but perhaps using the evaluated initializer in this situation would be slightly more consistent. After all, the initializer is still required for this to work - so the evaluation happens on the grounds that the expression that can be evaluated is there. It's not like this can be used as part of the evaluation: `declare const str: 'foo'`;

It's also worth noting that this code here should be expanded with support for enum members.